### PR TITLE
Update Writing plugin and Plugin pattern page

### LINKED ISSUE #2172: Update `writing a plugin` page
The current [writing a plugin](https://webpack.js.org/contribute/writing-a-plugin/) and [plugin patterns](https://webpack.js.org/contribute/plugin-patterns/) page is out of date.. It uses plugin method, which is deprecated. Moreover it doesn't uses ES6 classes, now that webpack has dropped support for node 4, we can safely use classes. I'll send PR, once we're sure, that we are good to use ES6 classes or not.

### PR DESCRIPTION
**Pull request by nveenjain**
Fixes #2172, Updated writing plugins and plugin patterns page to include ES6 classes (as ES6 is supported by webpack completely) and updates related to new tapable library.

### CODE REVIEW

**Comment by dhruvdutt** [comment]
Related to #2202.

File: src/content/contribute/writing-a-plugin.md [hunk 1]
```diff
@@ -133,76 +154,106 @@ Once we can latch onto the webpack compiler and each individual compilations, th
...
+For example:-
+
+```javascript
+this.hooks = {
+  shouldEmit: new SyncBailHook(["compilation"])
+};
+```
 
-`applyPlugins(name: string, args: any...)`
+It represents that the only hook supported is shouldEmit which is a hook of `SyncBailHook` type and the only parameter which will be passed to any plugin that taps into shouldEmit hook is compilation.
```
Comment (EugeneHlushko) [inline src/content/contribute/writing-a-plugin.md]: \`compilation\`

Comment (EugeneHlushko) [reply src/content/contribute/writing-a-plugin.md]: \`shouldEmit\`

File: src/content/contribute/writing-a-plugin.md [hunk 2]
```diff
@@ -133,76 +154,106 @@ Once we can latch onto the webpack compiler and each individual compilations, th
...
+  }
+}
 
 module.exports = FileListPlugin;
 ```
 
 ## Different Plugin Shapes
 
-A plugin can be classified into types based on the event it is registered to. Every event hook decides how it is going to apply the plugins in its registry.
+A plugin can be classified into types based on the event hooks it taps into. Every event hook is pre-defined as synchronous/asynchronous/waterfall/parallel and hook is called internally using call/callAsync method. The list of hooks that are supported/can be tapped into are generally specified in this.hooks property.
```
Comment (EugeneHlushko) [inline src/content/contribute/writing-a-plugin.md]: supported or can be tapped...

I would recommend to use or and comma separation here for the  `/` synchronous/asynchronous/waterfall/parallel

File: src/content/contribute/plugin-patterns.md [hunk 3]
```diff
@@ -10,63 +12,66 @@ Plugins grant unlimited opportunity to perform customizations within the webpack
...
-    // Explore each chunk (build output):
-    compilation.chunks.forEach(function(chunk) {
-      // Explore each module within the chunk (built inputs):
-      chunk.modules.forEach(function(module) {
-        // Explore each source file path that was included into the module:
-        module.fileDependencies.forEach(function(filepath) {
-          // we've learned a lot about the source structure now...
+class MyPlugin {
+  apply(compiler) {
+    compiler.hooks.emit.tapAsync("MyPlugin", (compilation, callback) => {
```
Comment (EugeneHlushko) [inline src/content/contribute/plugin-patterns.md]: we use single quotes for \`\`\`javascript i wont mark all of the places where you changed it or added as `"`, please change everywhere

File: src/content/contribute/plugin-patterns.md [hunk 4]
```diff
@@ -10,63 +12,66 @@ Plugins grant unlimited opportunity to perform customizations within the webpack
...
+      // Explore each chunk (build output):
+      compilation.chunks.forEach(function(chunk) {
```
Comment (EugeneHlushko) [inline src/content/contribute/plugin-patterns.md]: we use arrow functions in webpack/webpack codebase in such blocks, i would recommend changing these into arrow functions for style consistency

File: src/content/contribute/plugin-patterns.md [hunk 5]
```diff
@@ -77,22 +82,21 @@ You may also feed new file paths into the watch graph to receive compilation tri
...
-MyPlugin.prototype.apply = function(compiler) {
-  compiler.plugin('emit', function(compilation, callback) {
-
-    var changedChunks = compilation.chunks.filter(function(chunk) {
-      var oldVersion = this.chunkVersions[chunk.name];
-      this.chunkVersions[chunk.name] = chunk.hash;
-      return chunk.hash !== oldVersion;
-    }.bind(this));
-
+class MyPlugin{
```
Comment (EugeneHlushko) [inline src/content/contribute/plugin-patterns.md]: the formatting is off, spaces are missing

File: src/content/contribute/writing-a-plugin.md [hunk 6]
```diff
@@ -106,21 +101,47 @@ For more information on what callbacks are available on the `compiler`, `compila
 
 ## Async compilation plugins
 
-Some compilation plugin steps are asynchronous, and pass a callback function that _must_ be invoked when your plugin is finished running.
+Some plugin hooks are asynchronous. To tap into them, we can use `tap` method which will behave in synchronous manner or use one of `tapAsync` method or `tapPromise` method which are asyncronous methods.
+
+### tapAsync
+
+When we use tapAsync method to tap into plugins, we need to call the callback function which is supplied as the last argument to our function.
```
Comment (EugeneHlushko) [inline src/content/contribute/writing-a-plugin.md]: \`tapAsync\`

File: src/content/contribute/writing-a-plugin.md [hunk 7]
```diff
@@ -106,21 +101,47 @@ For more information on what callbacks are available on the `compiler`, `compila
...
 
-    // Do something async...
-    setTimeout(function() {
-      console.log("Done with async work...");
-      callback();
-    }, 1000);
+#### tapPromise
 
-  });
+When we use tapPromise method to tap into plugins, we need to return a promise which resolves when our asynchronous task is completed.
```
Comment (EugeneHlushko) [inline src/content/contribute/writing-a-plugin.md]: \`tapPromise\` and in other places too pls

File: src/content/contribute/writing-a-plugin.md [hunk 8]
```diff
@@ -133,76 +154,106 @@ Once we can latch onto the webpack compiler and each individual compilations, th
...
 
-`applyPluginsBailResult(name: string, args: any...)`
+Various types of hooks supported are :-
 
-This means that each of the plugin callbacks will be invoked one after the other with the specific `args`.
-This is the simplest format for a plugin. Many useful events like `"compile"`, `"this-compilation"` expect plugins to have synchronous execution.
+### Synchronous Hooks
 
-- __waterfall__ Plugins applied using
+- **SyncHook**
```
Comment (EugeneHlushko) [inline src/content/contribute/writing-a-plugin.md]: we use `__` for boldness, please fix this and next occurrences

File: src/content/contribute/writing-a-plugin.md [hunk 9]
```diff
@@ -133,76 +154,106 @@ Once we can latch onto the webpack compiler and each individual compilations, th
...
 
-`applyPluginsWaterfall(name: string, init: any, args: any...)`
+- **Bail Hooks**
+  - defined using `SyncBailHook[params]`
+  - Tapped into using `tap` method.
+  - Called using `call(...params)` method.
 
-Here each of the plugins are called one after the other with the args from the return value of the previous plugin. The plugin must take the order of its execution into account.
-It must accept arguments from the previous plugin that was executed. The value for the first plugin is `init`. This pattern is used in the Tapable instances which are related to the `webpack` templates like `ModuleTemplate`, `ChunkTemplate` etc.
+  In these type of hooks, each of the plugin callbacks will be invoked one after the other with the specific `args`. If any value is returned except undefined by any plugin, then that value is returned by hook and no further plugin callback is invoked. Many useful events like `"optimizeChunks"`, `"optimizeChunkModules"` are SyncBailHooks.
```
Comment (EugeneHlushko) [inline src/content/contribute/writing-a-plugin.md]: You dont need to double quote the events, just \`optimizeChunks\`

File: src/content/contribute/writing-a-plugin.md [hunk 10]
```diff
@@ -133,76 +154,106 @@ Once we can latch onto the webpack compiler and each individual compilations, th
...
 
-- __asynchronous__ When all the plugins are applied asynchronously using
+- **Waterfall Hooks**
+  - defined using `SyncWaterfallHook[params]`
+  - Tapped into using `tap` method.
+  - Called using `call( ... params)` method
 
-`applyPluginsAsync(name: string, args: any..., callback: (err?: Error) -> void)`
+  Here each of the plugins are called one after the other with the args from the return value of the previous plugin. The plugin must take the order of its execution into account.
```
Comment (EugeneHlushko) [inline src/content/contribute/writing-a-plugin.md]: args -> arguments

File: src/content/contribute/writing-a-plugin.md [hunk 11]
```diff
@@ -133,76 +154,106 @@ Once we can latch onto the webpack compiler and each individual compilations, th
...
+  It must accept arguments from the previous plugin that was executed. The value for the first plugin is `init`. Hence atleast 1 param must be supplied for waterfall Hooks. This pattern is used in the Tapable instances which are related to the `webpack` templates like `ModuleTemplate`, `ChunkTemplate` etc.
```
Comment (EugeneHlushko) [inline src/content/contribute/writing-a-plugin.md]: Hooks -> hooks

Comment (EugeneHlushko) [reply src/content/contribute/writing-a-plugin.md]: \`webpack\` -> webpack

File: src/content/contribute/writing-a-plugin.md [hunk 12]
```diff
@@ -133,76 +154,106 @@ Once we can latch onto the webpack compiler and each individual compilations, th
...
+### Asynchronous Hooks
 
-- __async waterfall__ The plugins will be applied asynchronously in the waterfall manner.
+- **Async Series Hook**
+  - defined using `AsyncSeriesHook[params]`
+  - Tapped into using `tap`/`tapAsync`/`tapPromise` method.
+  - Called using `callAsync( ... params)` method
 
-`applyPluginsAsyncWaterfall(name: string, init: any, callback: (err: Error, result: any) -> void)`
+  The plugin handler functions are called with all args and a callback function with the signature `(err?: Error) -> void`. The handler functions are called in order of registration. `callback` is called after all the handlers are called.
```
Comment (EugeneHlushko) [inline src/content/contribute/writing-a-plugin.md]: args -> arguments

**Comment by montogeek** [comment]
Thanks!

**Pull request merged**